api/extension-guides/chat.md

```diff
@@ -0,0 +1,204 @@
+stream.markdown('myStack.push(2); // adding another fish (number 2) \n');
+stream.markdown('console.log(myStack.pop()); // eating the top fish, will output: 2 \n')
+stream.markdown('``` \n')
+stream.markdown('So remember, Code Kitten, in a stack, the last fish in is the first fish out - which we tech cats call LIFO (Last In, First Out).');
+
+stream.button({
+    command: 'cat.meow',
+    title: vscode.l10n.t('Meow!'),
+    arguments: []
+});
+```
+
+In practice, extensions will send a request to the Language Model, and once they get a response from the Language Model they might process it, and decide if they should stream anything back to the user. VS Code Chat API is streaming based, and thus it is compatible with streaming Language Model APIs. This allows extensions to report progress and results continuously with the goal of having a smooth user experience.
+
+
+### Follow-ups
+
+After each request, followup providers are invoked to get suggested followup questions to show the user. The user can click the followup to send it to the chat. Followups are options for the user on how to take the conversation further.
+
+```typescript
+cat.followupProvider = {
+    provideFollowups(result: ICatChatResult, token: vscode.CancellationToken) {
+        if (result.metadata.command === 'teach') {
+            return [{
+                prompt: 'let us play',
+                title: vscode.l10n.t('Play with the cat')
+            } satisfies vscode.ChatFollowup];
+        }
+    }
+};
+```
+
+
+> **Tip:** Follow-ups should be written as questions or directions, not just concise commands.
+
+## Variables
+
+Orthogonal to chat participants, extensions can contribute `#variables`. Variables start with the `#` syntax and are resolved by an extension that contributes that variable or by VS Code if the variable of built-in (e.g `#file`, `#selection`...). Users explicitly include variables in their prompts, by pressing `#` VS Code will offer a list of registered variables with their description:
+
+![List of variables in chat](images/chat/variables.png)
+
+For example, a C++ extension might contribute a variable `#cpp_context` that would get resolved based on the state of the language service - what C++ version is being used and what C++ programming approach is preferred.
+
+Chat participants should not talk directly to each other, instead extensions can define an API that another chat participant extension depends on. Also, variables are resolved independent of the active chat participant, and thus they can be used as a mechanism to share context between participants. For example, `@workspace` already maintains an index of the current workspace. `@workspace` could contribute a variable `#codebase`, that would allow other extensions to leverage the codebase context in their prompts.
+
+Variable resolvers can offer multiple length levels for the variable value and VS Code will use the one depending on how many tokens are left in an LLM prompt.
```

Comment:
This section is not clear to me, not being too familiar with tokens and LLMs in general.

Reply:
This is removed. SHould be better now.